server/src/main/java/org/opensearch/node/Node.java

```diff
@@ -548,11 +549,17 @@ protected Node(
              */
             this.environment = new Environment(settings, initialEnvironment.configDir(), Node.NODE_LOCAL_STORAGE_SETTING.get(settings));
             Environment.assertEquivalent(initialEnvironment, this.environment);
-            if (DiscoveryNode.isSearchNode(settings) == false) {
-                nodeEnvironment = new NodeEnvironment(tmpSettings, environment);
-            } else {
-                nodeEnvironment = new NodeEnvironment(settings, environment, new FileCacheCleaner(this::fileCache));
-            }
+            IndexStoreListener.CompositeIndexStoreListener compositeIndexStoreListener = new IndexStoreListener.CompositeIndexStoreListener(
+                Stream.concat(
+                    pluginsService.filterPlugins(IndexStorePlugin.class)
+                        .stream()
+                        .map(IndexStorePlugin::getIndexStoreListener)
+                        .filter(Optional::isPresent)
+                        .map(Optional::get),
+                    Stream.of(new FileCacheCleaner(this::fileCache))
```

Comment:
`FileCacheCleaner` asserts that `FileCache` should not be `null`. `FileCache` is only initialized on searchNode and now with creating `FileCacheCleaner` on every node the assertion is not valid. We can keep the existing behavior of adding the FileCache listener only for search node and when `FileCache` is initialized on all the nodes (not only on search node) then this could be changed. Thoughts ?

Reply:
Since the assertion is wrapped by a `if (indexSettings.isRemoteSnapshot()) {` I don't think we are actually breaking the assertion, however I do agree with trying to keep the existing flow as close to same as possible and only adding `FileCacheCleaner` to the list of plugin implementations in case of search node sounds like a good change.